docs/proposals/19-migrating-to-prometheus-operator.md.md

```diff
@@ -0,0 +1,444 @@
+              cluster: shoot--project--name
+          replicas: 1
+          retention: 30d
+          routePrefix: /
+          ruleSelector:
+            matchLabels:
+              cluster: shoot--project--name
+          serviceAccountName: prometheus
+          serviceMonitorSelector:
+            matchLabels:
+              cluster: shoot--project--name
+          storage:
+            volumeClaimTemplate:
+              spec:
+                accessModes:
+                - ReadWriteOnce
+                resources:
+                  requests:
+                    storage: 20Gi
+          version: v2.35.0
+        ```
+
+        - In addition to a Prometheus, the configuration must also be created. To
+          do this, each `job` in the prometheus configuration will need to be
+          replaced with either a `ServiceMonitor`, `PodMonitor`, or `Probe`.
+          This `ServiceMonitor` would be picked up by the prometheus defined in
+          the previous step and scrape any service in the namespace
+          `shoot--project--name` that has the label `app=prometheus` on the port
+          called `metrics`.
+
+        ```yaml
+        apiVersion: monitoring.coreos.com/v1
+        kind: ServiceMonitor
+        metadata:
+          labels:
+            cluster: shoot--project--name
+          name: prometheus-job
+          namespace: shoot--project--name
+        spec:
+          endpoints:
+          - port: metrics
+          namespaceSelector:
+            matchNames:
+            - shoot--project--name
+          selector:
+            matchLabels:
+              app: prometheus
+        ```
+
+    1. The contract between `Prometheus` and its configuration. `Prometheus` can
+       discover different `*Monitors` using different labels. This is defined by
+       `matchLabels` for the corresponding configuration object. The
+       configuration objects (for example `ServiceMonitors`) should have the
+       label `cluster=shoot--project--name` and `Prometheus` will be configured
+       to discover these objects.
+
+    ```yaml
+    serviceMonitorSelector:
+      matchLabels:
+        cluster: shoot--project--name
+    ```
+
+    1. Prometheus needs to discover targets running in the shoot cluster.
+       Normally, this is done by changing the `api_server` field in the config
+       ([example][apiserver-example]). This is currently not possible with the
+       prometheus operator; however, there is an open [issue][prom-op-issue].
+
+        - Alternative 1: Use [additional scrape config]. In this case, the
+          prometheus config snippet is put into a secret and the
+          [prometheus-operator] will append it to the config. The downside here is
+          that it is only possible to have 1 `additional-scrape-config` per
+          Prometheus. This could be an issue if multiple components will need to
+          use this.
+
+        - Alternative 2: A second prometheus can be created that is running in
```

Comment:
I like this option. One question to it:
Would this agent instance also pick up the usual prometheus-operator CRDs?
If so, this option sounds like the best one to me.

Reply:
>Would this agent instance also pick up the usual prometheus-operator CRDs?

It will pick up whatever we configure it to pick up :) 

The critical aspect is that the "normal" Prometheus cannot discover any objects in the shoot because we cannot change the `api_server` field in `*Monitor` objects, but only in `Prometheus` objects. This means that the "normal" Prometheus can only scrape targets in the seed. The agent will have the `api_server` field set to the shoot's API Server and can do service discovery there. The agent will only scrape shoot targets and ignore everything in the seed.

Reply:
I will leave all alternatives, but make `alternative 2` the preferred approach.